File: Acwing/First_Level_0_Basic_Algorithm/Second_DataStructure/ds_3302_expression_evaluation/Main.java
package First_Level_0_Basic_Algorithm.Second_DataStructure.ds_3302_expression_evaluation;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

public class Main {
 
    static Stack<Character> op = new Stack<>();
    static Stack<Integer> n = new Stack<>();
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        char [] str = br.readLine().strip().toCharArray();
        int len = str.length;
        Map<Character,Integer> pr = new HashMap<Character,Integer>(){{
            put('+', 1);
            put('-', 1);
            put('*', 2); 
            put('/', 2);
        }};
        for(int i = 0 ; i < len ; ++i){
            char c = str[i];
            if(Character.isDigit(c)){
                int num = str[i] - '0';
                int j = i + 1;
                while(j < len && Character.isDigit(str[j]) ){
                    num = num * 10 + str[j] - '0';
                    j++;
                }
                i = j - 1;
                n.push(num);
            }else{
                if(c == '(') op.push(c);
                else if( c == ')'){
                    while(op.peek() != '(') calculate();
                    op.pop();
                }else{
                    while(!op.empty() && op.peek() != '('&& pr.get(op.peek()) >= pr.get(c))  calculate();
                    op.push(c);
                }
            }
        }
        br.close();
        while(op.size() > 0) calculate();
        System.out.println(n.pop());
    }

    public static void calculate(){
        int b = n.peek(); n.pop();
        int a = n.peek(); n.pop();
        char ops = op.peek(); op.pop();
        int res = 0;
        switch(ops){
            case '+': res = a + b ; break;
            case '-': res = a - b ; break;
            case '*': res = a * b ; break;
            case '/': res = a / b ; break;
        }
        n.push(res);
    }
}

File: Acwing/First_Level_0_Basic_Algorithm/Second_DataStructure/ds_3302_expression_evaluation/Main.cpp
#include<iostream>
#include<stack>
#include<algorithm>
#include<unordered_map>
using namespace std;
const int N = 100010;
stack<char> op;
stack<int> n;
char str[N];

void calculate(){
    int res = 0;
    auto a = n.top(); n.pop();
    auto b = n.top(); n.pop();
    char ops = op.top(); op.pop();
    if(ops == '+') res = a + b;
    else if(ops == '-') res = b - a;
    else if(ops == '*') res = a * b;
    else res = b / a;
    n.push(res);
}

int main(){
    unordered_map<char, int> pr{{'+', 1}, {'-', 1}, {'*', 2}, {'/', 2}};
    cin >> str;
    for(int i = 0 ; str[i] ; i++){
        char c = str[i];
        if(isdigit(c)){
            int num = c - '0';
            int j = i + 1;
            while(isdigit(str[j])){
                num = num * 10 + str[j++] - '0';
            }
            i = j - 1;
            n.push(num);
        }else{  
            if(c == '('){
                op.push(c);
            }else if( c == ')'){
                while(op.top() != '(') calculate();
                op.pop();
            }else{
                while(op.size() && op.top() != '(' && pr[c] <= pr[op.top()]) calculate();
                op.push(c);
            }
        }
    }
    while (op.size()) calculate();
    cout << n.top() << endl;
}